crates/cargo-util/src/paths.rs

```diff
@@ -185,10 +185,45 @@ pub fn write<P: AsRef<Path>, C: AsRef<[u8]>>(path: P, contents: C) -> Result<()>
 /// write_atomic uses tempfile::persist to accomplish atomic writes.
 pub fn write_atomic<P: AsRef<Path>, C: AsRef<[u8]>>(path: P, contents: C) -> Result<()> {
     let path = path.as_ref();
-    let mut tmp = TempFileBuilder::new()
-        .prefix(path.file_name().unwrap())
-        .tempfile_in(path.parent().unwrap())?;
+    let mut builder = TempFileBuilder::new();
+
+    builder.prefix(path.file_name().unwrap());
+
+    // On unix platforms, use the same permissions as the original file. Copy only the
+    // user/group/other read/write/execute permission bits. While we need to also set the
+    // permissions again later to bypass the umask, we still need to set the permissions here as
+    // well so that we don't create a more-permissive file than the original. The tempfile lib says
```

Comment:
The default seems to be [0o600](https://docs.rs/tempfile/3/tempfile/struct.Builder.html#method.permissions), which is pretty safe I believe.  Do we really need to set permission for tempfiles?
Granted, if the `Cargo.toml` was readonly, maybe `cargo add` should fail.

Reply:
I can remove that part and just use the default 600 permissions to simplify it. tempfile could change their default in the future, but that's probably unlikely.

Reply:
Yeah that could simplify the logic a bit.

Reply:
It now only sets the permissions once after the file has been created. The file will be created with tempfile's default of 600.